### PR TITLE
Update unsafe class defining to work on Java 17+

### PR DESCRIPTION
**Pull request by mcculls**
Java 17 has removed `Unsafe.defineAnonymousClass` in favour of `Lookup.defineHiddenClass`

### CODE REVIEW

File: core/test/com/google/inject/MethodInterceptionTest.java [hunk 1]
```diff
@@ -269,6 +270,11 @@ protected void configure() {
 
   @Test
   public void testNotInterceptedMethodsInInterceptedClassDontAddFrames() {
+    // Test relies on Enhancer appearing in stack traces which isn't true for Java17 hidden classes
```
Comment (sameb) [inline core/test/com/google/inject/MethodInterceptionTest.java:273]: can we scope this around the ==true portion, so that the ==false portion is still tested?

alternately, can we add a java17+ test that the hidden $Enhancer classes never appear in stack traces so that the same property is tested (but perhaps in a broader manner now -- incl that intercepted methods don't get it) ?

same thing w/ the FastClass stacks below.

**Review by sameb — COMMENTED** [review]
Thanks @mcculls!  In addition to the comment, would it be possible to add some kind of package-level description of the flow this code takes?  It's somewhat hard to follow today already, and the additional ClassDefiner impls makes it even harder.  Some kind of README.md or comments in an entry-point class that describe how the classes work together and when we fallback from A->B->C, etc.. would be helpful in explaining this.

Thanks!

File: core/test/com/google/inject/MethodInterceptionTest.java (hunk 1 above)
Comment (mcculls) [reply core/test/com/google/inject/MethodInterceptionTest.java]: Managed to fix this another way: 385c857e8
